### PR TITLE
add JWT authorization for API

### LINKED ISSUE #296: API Protection with Capcha
# Problem
Our ofDB-API is completly open and could be modified with any script.
That would cause a lot of work to detect the errors and reset the databasis.

# Solution
We should protect the ofDB with ~~Tokens~~ captchas for unauthenticated users.
~~Each frontend or script would need an authentification.~~
~~That is independet of the questions, if users have a login or not.~~

# ToDos

- [x] Describe captcha API
- [x] Implement captcha service
- [ ] Prepare the frontend
- [x] Restrict API
- [ ] Deploy dev backend
- [ ] Deploy dev frontend
- [ ] Deploy backend & frontend

### LINKED ISSUE #310: Limit write requests for unauthorized users
In addition to #308 we could limit write request by looking for the IP address.

### PR DESCRIPTION
**Pull request by magdaddy**

### CODE REVIEW

File: openfair.db [hunk 1]
```diff
@@ -0,0 +1 @@
+../ofdb-inhalt/openfair.db.normal
```
Comment (flosse) [inline openfair.db]: please remove this file!

File: ofdb-boundary/src/lib.rs [hunk 2]
```diff
@@ -1213,3 +1213,24 @@ impl From<Entry> for UpdatePlace {
         }
     }
 }
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "extra-derive", derive(Debug, Clone, PartialEq))]
+pub struct Error {
```
Comment (flosse) [inline ofdb-boundary/src/lib.rs]: I think this `Error` struct should not be part of this PR.

File: Cargo.toml [hunk 3]
```diff
@@ -71,10 +74,14 @@ ofdb-entities = { version = "*", features = ["builders"] }
 walkdir = "2"
 
 [features]
-default = ["frontend", "email", "clearance"]
+default = ["frontend", "email", "clearance", "cookies", "jwt"]
+
+app = []
```
Comment (flosse) [inline Cargo.toml]: what is this feature for?

File: src/ports/web/jwt.rs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+use std::{
+    collections::HashSet,
+    convert::TryFrom,
+    sync::{Mutex, MutexGuard},
+};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Claims {
+    /// The email in our case
+    sub: String,
```
Comment (flosse) [inline src/ports/web/jwt.rs:14]: is `sub`  a standard key for JWT? Or can we directly use `email`?

File: ofdb-boundary/src/lib.rs [hunk 5]
```diff
@@ -1213,3 +1213,24 @@ impl From<Entry> for UpdatePlace {
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub code: Option<u16>,
+    /// Error details
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub detail: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "extra-derive", derive(Debug, Clone))]
+pub struct JwtToken {
```
Comment (flosse) [inline ofdb-boundary/src/lib.rs:619]: Is this the common JWT format? Or can we just use `JwtToken(String)`?

**Review by flosse — CHANGES_REQUESTED** [review]

File: openfair.db (hunk 1 above)
Comment (flosse) [reply openfair.db]: ... btw: you could add `openfair.db` to `.gitignore` :)

Comment (magdaddy) [reply openfair.db]: done

File: src/ports/web/jwt.rs (hunk 4 above)
Comment (magdaddy) [reply src/ports/web/jwt.rs:14]: exactly. This is the standard key for identifying the "Subject" for which the claims are issued, which is the email address in our case.

File: ofdb-boundary/src/lib.rs (hunk 2 above)
Comment (magdaddy) [reply ofdb-boundary/src/lib.rs]: yeah, a rebasing mistake. done.

**Comment by flosse** [comment]
please squash the two commits into one, thx!

**Review by flosse — APPROVED** [review]

**Comment by flosse** [comment]
@magdaddy Would you mind to extend the `openapi.yaml` to describe the usage?

**Comment by magdaddy** [comment]
openapi.yaml is updated now.

**Review by flosse — APPROVED** [review]

**Comment by magdaddy** [comment]
I'm very satisfied with the guard redesign now. I think it's a good direction to take for the future. Looking forward to your opinion.

File: src/ports/web/guards.rs [hunk 6]
```diff
@@ -1,127 +1,159 @@
...
+    bearer_tokens: Vec<String>,
+    account_email: Option<String>,
+    has_captcha: bool,
+}
 
-impl Account {
-    pub fn email(&self) -> &str {
-        &self.0
+impl Auth {
+    // pub fn unauthorized() -> AppError {
```
Comment (flosse) [inline src/ports/web/guards.rs:37]: can we remove this?

**Review by flosse — APPROVED** [review]

**Comment by flosse** [comment]
> I'm very satisfied with the guard redesign now. I think it's a good direction to take for the future. Looking forward to your opinion.

Thanks a lot!
LGTM

File: src/ports/web/guards.rs (hunk 6 above)
Comment (magdaddy) [reply src/ports/web/guards.rs:37]: yeah, the unauthorized() method is not needed, forgot to remove it.

The simple `Account` implementation based on `Auth` was done because the frontend code depends on it a lot. As it's going to be removed anyway soon, I didn't bother to change everything there. When the frontend is replaced by seed app, `Account` can be removed.

**Comment by wellemut** [comment]
> 
> 
> openapi.yaml is updated now.
@magdaddy I somehow can not find, where it is documented and how somebody could integrate it in his/her frontend.

- Could you give a direct link or post a screenshot here?
- Is it possible to use a token in your API instead of a capcha?

Thank you very much. 
As soon as you have cleared that, I will notify all API-Users: https://github.com/kartevonmorgen/openfairdb/issues/299

**Comment by magdaddy** [comment]
@wellemut  Well, it's in the openapi.yaml... and in the end there isn't much to document.

It's simply an additional way of authentication/authorization. Instead of implicitly depending on cookies, you can log in and explicitly get a token back. This is valid for 24h, and you simply pass it in the HTTP header (Authorization: Bearer <token>) for any requests that require authorization.

It is an **alternative for cookies**, because cookies are a bit awkward in context of SPA's and other consumers that are not webbrowsers (e.g. scripts). For example, the React Frontend never knows if it is currently logged in (because it can't access the cookies).

It is **not** an alternative for captchas (because you need to have an account), and also not for the tokens of organizations.

**Comment by wellemut** [comment]
@navid-kalaei 
The https://api.ofdb.io stills runs on the old system.

But the https://dev.ofdb.io has allready the new authentication System if you want to create or edit places. (Reading is still working just like this.

You have the Following options:

1. bearerAuth  (http, Bearer) 
2. captchaCookieAuth  (apiKey)  (For anonymous users)
3. jwtAuth  (http, Bearer) (For Scripts and Organisational Plattforms which have a fixed token)
4. userEmailCookieAuth  (apiKey)  (For Users with an account)

Please check out the yaml for further details:

-    go to https://editor.swagger.io
-   go to File -> import URL
-    enter https://raw.githubusercontent.com/kartevonmorgen/openfairdb/master/openapi.yaml
-    enjoy ;-)


Value:
grafik


Captcha
grafik

**Comment by wellemut** [comment]
Navid will use this system https://jwt.io/
